Sporadic unit test failure: StoreClientFactoryTest.testZkSessionExpiryRetry

**Problem description**
```
io.pravega.controller.store.client.StoreClientFactoryTest > testZkSessionExpiryRetry FAILED
    java.lang.RuntimeException: Exception of the wrong type. Was expecting class org.apache.zookeeper.KeeperException$SessionExpiredException but got: org.apache.zookeeper.KeeperException$NoNodeException
        at io.pravega.test.common.AssertExtensions.assertThrows(AssertExtensions.java:75)
        at io.pravega.controller.store.client.StoreClientFactoryTest.testZkSessionExpiryRetry(StoreClientFactoryTest.java:84)
        Caused by:
        org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /test/test
            at org.apache.zookeeper.KeeperException.create(KeeperException.java:115)
            at org.apache.zookeeper.KeeperException.create(KeeperException.java:51)
            at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:2019)
            at org.apache.curator.framework.imps.GetDataBuilderImpl$4.call(GetDataBuilderImpl.java:327)
            at org.apache.curator.framework.imps.GetDataBuilderImpl$4.call(GetDataBuilderImpl.java:316)
            at org.apache.curator.connection.StandardConnectionHandlingPolicy.callWithRetry(StandardConnectionHandlingPolicy.java:64)
            at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:100)
            at org.apache.curator.framework.imps.GetDataBuilderImpl.pathInForeground(GetDataBuilderImpl.java:313)
            at org.apache.curator.framework.imps.GetDataBuilderImpl.forPath(GetDataBuilderImpl.java:304)
            at org.apache.curator.framework.imps.GetDataBuilderImpl.forPath(GetDataBuilderImpl.java:35)
            at io.pravega.controller.store.client.StoreClientFactoryTest.lambda$testZkSessionExpiryRetry$4(StoreClientFactoryTest.java:84)
            at io.pravega.test.common.AssertExtensions.assertThrows(AssertExtensions.java:64)
            ... 1 more
```

**Problem location**
Controller.

**Suggestions for an improvement**
Fix unit test or underlying code.

Observed assertion failure:

```
io.pravega.controller.store.client.StoreClientFactoryTest > testZkSessionExpiryRetry FAILED
    java.lang.AssertionError: expected:<3> but was:<4>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at io.pravega.controller.store.client.StoreClientFactoryTest.testZkSessionExpiryRetry(StoreClientFactoryTest.java:90)
```

Found another instance of this failure:
```
io.pravega.controller.store.client.StoreClientFactoryTest > testZkSessionExpiryRetry FAILED
    java.lang.AssertionError: expected:<3> but was:<4>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at io.pravega.controller.store.client.StoreClientFactoryTest.testZkSessionExpiryRetry(StoreClientFactoryTest.java:90)
```

--- a/controller/src/test/java/io/pravega/controller/store/client/StoreClientFactoryTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/client/StoreClientFactoryTest.java
@@ -19,6 +19,7 @@ import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.state.ConnectionState;
 import org.apache.curator.test.TestingServer;
 import org.apache.zookeeper.KeeperException;
 import org.junit.After;
@@ -27,6 +28,7 @@ import org.junit.Test;
 
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.assertFalse;
+
 
 public class StoreClientFactoryTest extends ThreadPooledTestSuite {
     TestingServer zkServer;
@@ -59,17 +61,17 @@ public class StoreClientFactoryTest extends ThreadPooledTestSuite {
             return !sessionExpiry.isDone();
         };
         Consumer<Void> expirationHandler = x -> sessionExpiry.complete(null);
-
         CuratorFramework client = StoreClientFactory.createZKClient(ZKClientConfigImpl.builder().connectionString(zkServer.getConnectString())
-                .namespace("test").maxRetries(10).initialSleepInterval(10).secureConnectionToZooKeeper(false).sessionTimeoutMs(15000).build(),
+                        .namespace("test").maxRetries(10).initialSleepInterval(10).secureConnectionToZooKeeper(false).sessionTimeoutMs(15000).build(),
                 canRetrySupplier, expirationHandler);
-        
+
         client.getZookeeperClient().getZooKeeper().getTestable().injectSessionExpiration();
-        
+
         sessionExpiry.join();
 
-        assertFalse(client.getZookeeperClient().getZooKeeper().getState().isConnected());
-
+        // verify that we fail with session expiry and we fail without retrying.
+        AssertExtensions.assertEventuallyThrows(KeeperException.SessionExpiredException.class, () -> client.getData().forPath("/test"),
+                2000, 30000L);
         assertTrue(expirationRetryCounter.get() > 0);
     }
 

--- a/controller/src/test/java/io/pravega/controller/store/client/StoreClientFactoryTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/client/StoreClientFactoryTest.java
@@ -19,16 +19,11 @@ import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 import org.apache.curator.framework.CuratorFramework;
-import org.apache.curator.framework.state.ConnectionState;
 import org.apache.curator.test.TestingServer;
 import org.apache.zookeeper.KeeperException;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
-
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.assertFalse;
-
 
 public class StoreClientFactoryTest extends ThreadPooledTestSuite {
     TestingServer zkServer;
@@ -46,33 +41,6 @@ public class StoreClientFactoryTest extends ThreadPooledTestSuite {
     @After
     public void tearDown() throws IOException {
         zkServer.stop();
-    }
-    
-    @Test(timeout = 60000)
-    public void testZkSessionExpiryRetry() throws Exception {
-        CompletableFuture<Void> sessionExpiry = new CompletableFuture<>();
-        AtomicInteger expirationRetryCounter = new AtomicInteger();
-
-        Supplier<Boolean> canRetrySupplier = () -> {
-            if (sessionExpiry.isDone()) {
-                expirationRetryCounter.incrementAndGet();
-            }
-
-            return !sessionExpiry.isDone();
-        };
-        Consumer<Void> expirationHandler = x -> sessionExpiry.complete(null);
-        CuratorFramework client = StoreClientFactory.createZKClient(ZKClientConfigImpl.builder().connectionString(zkServer.getConnectString())
-                        .namespace("test").maxRetries(10).initialSleepInterval(10).secureConnectionToZooKeeper(false).sessionTimeoutMs(15000).build(),
-                canRetrySupplier, expirationHandler);
-
-        client.getZookeeperClient().getZooKeeper().getTestable().injectSessionExpiration();
-
-        sessionExpiry.join();
-
-        // verify that we fail with session expiry and we fail without retrying.
-        AssertExtensions.assertEventuallyThrows(KeeperException.SessionExpiredException.class, () -> client.getData().forPath("/test"),
-                2000, 30000L);
-        assertTrue(expirationRetryCounter.get() > 0);
     }
 
     /**

--- a/controller/src/test/java/io/pravega/controller/store/client/StoreClientFactoryTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/client/StoreClientFactoryTest.java
@@ -26,6 +26,7 @@ import org.junit.Before;
 import org.junit.Test;
 
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertFalse;
 
 public class StoreClientFactoryTest extends ThreadPooledTestSuite {
     TestingServer zkServer;
@@ -67,9 +68,8 @@ public class StoreClientFactoryTest extends ThreadPooledTestSuite {
         
         sessionExpiry.join();
 
-        // verify that we fail with session expiry and we fail without retrying.
-        AssertExtensions.assertEventuallyThrows(KeeperException.SessionExpiredException.class, () -> client.getData().forPath("/test"),
-                2000, 30000L);
+        assertFalse(client.getZookeeperClient().getZooKeeper().getState().isConnected());
+
         assertTrue(expirationRetryCounter.get() > 0);
     }
 

